packages/SwingSet/src/vats/timer/vat-timer.js

```diff
@@ -1,98 +1,854 @@
 // @ts-check
 
-import { Nat } from '@agoric/nat';
-import { assert, details as X } from '@agoric/assert';
+import { E } from '@endo/eventual-send';
 import { Far, passStyleOf } from '@endo/marshal';
-import { makeNotifierFromAsyncIterable } from '@agoric/notifier';
 import { makePromiseKit } from '@endo/promise-kit';
-
-import { makeTimedIterable } from './timed-iteration.js';
+import { Nat } from '@agoric/nat';
+import { assert } from '@agoric/assert';
+import {
+  provideKindHandle,
+  defineDurableKind,
+  defineDurableKindMulti,
+  makeScalarBigMapStore,
+  makeScalarBigWeakMapStore,
+} from '@agoric/vat-data';
+import { provide, makeLegacyWeakMap } from '@agoric/store';
 import { TimeMath } from './timeMath.js';
 
-export function buildRootObject(vatPowers) {
+// RAM usage: O(number of outstanding delay() promises) +
+// O(number of notifiers??). setWakeup() only consumes DB.
+
+/**
+ * // Handler is a Far object with .wake(time) used for callbacks
+ *
+ * @typedef Handler
+ * @property {(scheduled: Timestamp) => unknown} wake
+ * @typedef {unknown} CancelToken
+ * // CancelToken must be pass-by-reference and durable, either local or remote
+ * @typedef Event
+ * @property {() => void} scheduleYourself
+ * @property {() => void} fired
+ * @property {() => void} cancel
+ * @typedef {MapStore<TimestampValue, Event[]>} Schedule
+ * @typedef {WeakMapStore<CancelToken, Event[]>} CancelTable
+ * @typedef {unknown} PromiseEvent
+ * @typedef WakeupPromiseControls
+ * @property {(scheduled: Timestamp) => void} resolve
+ * @property {(err: unknown) => void} reject
+ * @typedef {LegacyWeakMap<PromiseEvent, WakeupPromiseControls>} WakeupPromiseTable
```

Comment:
Oh, it looks like I could store the resolve/reject functions in a Far (even though they're never revealed off-vat), and that would let us use a `WeakMapStore` instead of the `LegacyWeakMap`. I'll do that.

The reason we can'ut use a (virtual) `BigWeakMapStore` is that we don't have virtual/durable Promises. The #4567 "durable notifier" plan will need this pattern too, to hold both the Notifier's `Promise` and it's resolve/reject control functions.

Reply:
I see it's done. Resolving.